target/octeon: add Linux 6.6 support and set it as default

rsalvaterra opened the pull request:
A rather trivial addition. Patches automatically rebased. Also add the new
CONFIG_USB_DWC3_OCTEON (disabled) to the generic 6.6 config and sort it, while
at it.

The second commit removes support for Linux 6.1 and sets 6.6 as the default.

Run-tested: UniFi USG

Signed-off-by: Rui Salvaterra \<rsalvaterra@gmail.com\>

rsalvaterra committed:
octeon: add support for testing kernel 6.6

A rather trivial addition. Patches automatically rebased. Also add the new
CONFIG_USB_DWC3_OCTEON (disabled) to the generic 6.6 config and sort it, while
at it.

Signed-off-by: Rui Salvaterra <rsalvaterra@gmail.com>
Link: https://github.com/openwrt/openwrt/pull/15838
Signed-off-by: Robert Marko <robimarko@gmail.com>

dangowrt commented:
Looks like at least one new Kconfig symbol is still missing:
https://github.com/openwrt/openwrt/actions/runs/9731831236/job/26856786854?pr=15838#step:31:34

rsalvaterra commented:
> Looks like at least one new Kconfig symbol is still missing: https://github.com/openwrt/openwrt/actions/runs/9731831236/job/26856786854?pr=15838#step:31:34

Indeed, how odd, I wasn't prompted for it on `make kernel_oldconfig`. Will fix, thanks!

rsalvaterra commented:
Ok, so `CONFIG_USB_DWC3_OCTEON` is the new symbol, which depends on `CONFIG_USB_DWC3` being enabled. I don't understand why in the CI we have `CONFIG_USB_DWC3_OCTEON=m` , when in the pull request's `config-6.6` the symbol is disabled. What am I missing?

dangowrt commented on target/linux/octeon/patches-6.6/700-allocate_interface_by_label.patch:
```diff
@@ -0,0 +1,37 @@
+ 	int ret;
+ 
++	if (priv->of_node)
++		label = of_get_property(priv->of_node, "label", NULL);
```
Can we drop this whole patch maybe in favor of solving this in userspace, similar to https://github.com/openwrt/openwrt/pull/15593

rsalvaterra replied:
I tried. It's easily doable if the labels are are located in `soc@0/mdio@1180000001800/ethernet-phy@n` (in the specific case of the USG), but for all other devices they're located in `soc@0/pip@11800a0000000/interface@n/ethernet@n`. In the latter case, I couldn't find a trivial way to obtain the device which matched the label. In the former case, it's a script almost identical to yours. If anyone else would like to give it a shot, please do, but there's another elephant in the room (Itus Shield, see below).

dangowrt commented on target/linux/octeon/patches-6.6/100-mips_image_cmdline_hack.patch:
```diff
@@ -0,0 +1,38 @@
+ config SYNC_R4K
+ 	bool
+ 
++config IMAGE_CMDLINE_HACK
```
As we are using Device Tree on this platform, why do we still need this hack?
Doesn't `/chosen/bootargs` work just as fine?

rsalvaterra replied:
Oh, I wish. There are Octeon subtargets for which we don't have a device tree at all (e.g. the Itus Shield), it's inherited from the bootloader. Unless we have device trees for all devices, I'm not seeing a trivial way of getting rid of this hack. It's **extremely** desirable, however, since Octeon is the last target requiring this.

flygarn12 commented:
I am testing a initramfs image at this time, after 12minutes it still looks ok. 
Are we using the test kernel function now? 
When we got kernel 6.1 it was right away put as only kernel without testing option?

flygarn12 commented:
Have anyone tested this on edgerouter4 with a crashreboot after a couple of hours?

flygarn12 reviewed (changes requested):
Have run it over 24h on Edgerouter 4 including with LAN3 SFP module without problems. My recommendation is to do the same process as with 6.1 and skip the test kernel function and make 6.6 default kernel directly.

bluecmd commented:
Seems to work on my Cisco vEdge 1000 with the only difference that I noticed is this new error in dmesg:

```
[   36.576944] jitterentropy: Initialization failed with host not compliant with requirements: 9
[   36.682824] kmodloader: 1 module could not be probed
[   36.687869] kmodloader: - jitterentropy_rng - 0
```

Back on a unit with 6.1 I see that `jitterentropy_rng` is loaded. I do not think I care about this module, but it does strike me as a potential regression.

rsalvaterra committed:
octeon: set Linux 6.6 as the default kernel

And remove support for 6.1 while at it.

Signed-off-by: Rui Salvaterra <rsalvaterra@gmail.com>
Link: https://github.com/openwrt/openwrt/pull/15838
Signed-off-by: Robert Marko <robimarko@gmail.com>

rsalvaterra commented:
Pull request updated with the commit setting 6.6 as the default kernel and removing 6.1. Whilst I'd love to get rid of all the hacks this target requires (if at all possible!), the rabbit hole I've been into the last week tells me this will be a "project" in itself. Therefore, I believe that shouldn't be a blocker to merging this pull request.

rsalvaterra commented:
> Seems to work on my Cisco vEdge 1000 with the only difference that I noticed is this new error in dmesg:
> 
> ```
> [   36.576944] jitterentropy: Initialization failed with host not compliant with requirements: 9
> [   36.682824] kmodloader: 1 module could not be probed
> [   36.687869] kmodloader: - jitterentropy_rng - 0
> ```
> 
> Back on a unit with 6.1 I see that `jitterentropy_rng` is loaded. I do not think I care about this module, but it does strike me as a potential regression.

Seems to be working as intended. The Octeon's CPUs most likely don't generate enough jitter to be used as a reliable source of entropy.

bluecmd commented:
> Seems to be working as intended. The Octeon's CPUs most likely don't generate enough jitter to be used as a reliable source of entropy.

Seems reasonable, thanks for checking! :+1:

bluecmd commented:
> I don´t have a `jitterentropy_rng` line in either 6.1 or 6.6 for the ER4 in the boot log. The only lines that contain any form of "rng" are these lines:

Interesting. Do you have a `/etc/modules.d/09-crypto-rng` file? I pull in `crypto-rng` to get CryptoAPI which includes `jitterentropy_rng` without any way of disabling it, so I would assume this would be a nag on all Octeons with `crypto-rng` installed.

flygarn12 commented:
> Interesting. Do you have a `/etc/modules.d/09-crypto-rng` file? I pull in `crypto-rng` to get CryptoAPI (which happens to include `jitterentropy_rng` without any way of disabling it), so I would assume this would be a nag on all Octeons with `crypto-rng` installed.
In 6.1 the answer is no, I have `09-crypto-aead` and `09-crypto-null` in that folder.

klondi commented:
> > Seems to work on my Cisco vEdge 1000 with the only difference that I noticed is this new error in dmesg:
> > ```
> > [   36.576944] jitterentropy: Initialization failed with host not compliant with requirements: 9
> > [   36.682824] kmodloader: 1 module could not be probed
> > [   36.687869] kmodloader: - jitterentropy_rng - 0
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > Back on a unit with 6.1 I see that `jitterentropy_rng` is loaded. I do not think I care about this module, but it does strike me as a potential regression.
> 
> Seems to be working as intended. The Octeon's CPUs most likely don't generate enough jitter to be used as reliable sources of entropy.

Hi, that error means that the health checks for the entropy jitter RNG have failed. If this did not happen on 6.1 it might be due to one of three commits: https://github.com/torvalds/linux/commit/3fde2fe99aa6dacd4151c87382b07ce7f30f0a52 https://github.com/torvalds/linux/commit/d23659769ad1bf2cbafaa0efcbae20ef1a74f77e https://github.com/torvalds/linux/commit/bb897c55042e9330bcf88b4b13cbdd6f9fabdd5e

The Octeon processors use a 5 https://www.realworldtech.com/octeon/3/  (or in your case likely 8 https://old.hotchips.org/wp-content/uploads/hc_archives/hc23/HC23.18.1-manycore/HC23.18.110-OCTEON_II-Kessler-Cavium%20-%20Copy.pdf ) stage pipeline.

Given that the Octeon arch doesn't have a branch or call predictor and the Octeon2 has a relatively large one per core. That means that branch and call misspredicts don't play such a big role in adding jitter.

Given the relatively large cache and that SHA-3 has a 25 word state which can fit the processor's registers and that the memory allocated is only 2KB (a page taking the one TLB entry), I doubt that memory accesses are a significant source of jitter either.

That basically only leaves interrupts. Since OpenWRT by default does no IRQ balancing that means that if the CPU used to do the health check is not CPU 0 you will likely get no jitter from these either.

Octeon doesn't do frequency or voltage scaling either. Not that it would matter because the HPT used is the cycle counter which is available on each core.

So the result you see is kind of exepected. If all fits into L1 cache, there is a fairly stable pipeline where non memory  instructions always take the same time, the branch misspredict pattern is always the same, there are no interrupts, and the timer is a cycle counter, there is very little space left for jitter to provide entropy. The health checks detect that and prevent this RNG from being used.

flygarn12 commented:
> > Interesting. Do you have a `/etc/modules.d/09-crypto-rng` file? I pull in `crypto-rng` to get CryptoAPI (which happens to include `jitterentropy_rng` without any way of disabling it), so I would assume this would be a nag on all Octeons with `crypto-rng` installed.
> 
> In 6.1 the answer is no, I have `09-crypto-aead` and `09-crypto-null` in that folder.
In 6.6 (initramfs) I don't have any 09 at all, for crypto I have `02-crypto-hash` and `04-crypto-crc32c` in the folder.

flygarn12 reviewed (approved):
Tested on ER4 and works as expected.

bluecmd reviewed (approved):
We've been running this PR for a while now and it works fine with our vEdges

robimarko commented:
Thanks! Rebased on top of main and merged!

robimarko commented:
BTW, as for the RNG error, I am also randomly hitting it on all platforms.
As far as I was able to narrow it down, it happens mostly in initramfs and seems to be caused by system time changing, because after NTP adjust the time and you reboot it magically works.
